toucan_connectors/one_drive/one_drive_connector.py

```diff
@@ -207,37 +273,37 @@ def _retrieve_data(self, data_source: OneDriveDataSource) -> pd.DataFrame:
         if not data_source.sheet and not data_source.table:
             raise ValueError('You must specify at least a sheet or a table')
 
-        df_all = pd.DataFrame()
+        file_names = None
+        if data_source.match:
+            file_names = self._retrieve_files_path(data_source)
 
-        workbook_elements_list = []
-        workbook_key_column = ''
-        if data_source.sheet:
-            workbook_elements_list = data_source.sheet
-            workbook_key_column = '__sheetname__'
-        else:
-            workbook_elements_list = data_source.table
-            workbook_key_column = '__tablename__'
-        workbook_elements_list = workbook_elements_list.replace(', ', ',').split(',')
+        df_all = pd.DataFrame()
+        workbook_elements_list, workbook_key_column = _prepare_workbook_elements(data_source)
 
         for workbook_element in workbook_elements_list:
-            url = self._format_url(data_source, workbook_element)
-
-            response = self._run_fetch(url)
-
-            data = response.get('values')
-
-            if not data:
-                logging.getLogger(__name__).debug('No data retrieved from response')
-                return pd.DataFrame()
+            if file_names:
+                urls = self._format_urls(data_source, workbook_element, file_names)
+            else:
+                urls = [self._format_url(data_source, workbook_element, data_source.file)]
 
-            cols = data[0]
-            data.pop(0)
+            data = []
+            for url in urls:
+                try:
+                    # TODO: some day make it async
+                    data.append(self._run_fetch(url).get('values'))
```

Comment:
if you're afraid of the time it could take... i can suggest something like..
```suggestion
            def url_yielder():
                for u in urls:
                    try:
                        yield self._run_fetch(u).get('values')
                    except requests.exceptions.HTTPError:
                         ...
            data = list(url_yielder())
```
...you can use `yield` and [`list-compression`](https://switowski.com/blog/for-loop-vs-list-comprehension#:~:text=List%20comprehensions%20are%20often%20not%20only%20more%20readable%20but%20also%20faster%20than%20using%20%E2%80%9Cfor%20loops.%E2%80%9D) there and then async in the future...

Reply:
A tiny step toward a more efficient solution. It won't be enough but I take it.